acir/src/native_types/witness.rs

```diff
@@ -27,23 +27,75 @@ impl Witness {
     pub const fn can_defer_constraint(&self) -> bool {
         true
     }
+}
+
+impl From<u32> for Witness {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+
+/// A map from the witnesses in a constraint system to the field element values
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Default, Serialize, Deserialize)]
+pub struct WitnessMap(BTreeMap<Witness, acir_field::FieldElement>);
+
+impl WitnessMap {
+    pub fn new() -> Self {
+        Self(BTreeMap::new())
+    }
+    pub fn get(&self, witness: &Witness) -> Option<&acir_field::FieldElement> {
+        self.0.get(witness)
+    }
+    pub fn get_index(&self, index: u32) -> Option<&acir_field::FieldElement> {
+        self.0.get(&index.into())
+    }
+    pub fn contains_key(&self, key: &Witness) -> bool {
+        self.0.contains_key(key)
+    }
+    pub fn insert(
+        &mut self,
+        key: Witness,
+        value: acir_field::FieldElement,
+    ) -> Option<acir_field::FieldElement> {
+        self.0.insert(key, value)
+    }
+}
 
-    pub fn to_bytes(
-        witnesses: &std::collections::BTreeMap<Witness, acir_field::FieldElement>,
-    ) -> Vec<u8> {
-        let buf = rmp_serde::to_vec(witnesses).unwrap();
+impl Index<&Witness> for WitnessMap {
+    type Output = acir_field::FieldElement;
+
+    fn index(&self, index: &Witness) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+impl From<BTreeMap<Witness, acir_field::FieldElement>> for WitnessMap {
+    fn from(value: BTreeMap<Witness, acir_field::FieldElement>) -> Self {
+        Self(value)
+    }
+}
+
+impl From<WitnessMap> for Vec<u8> {
+    fn from(val: WitnessMap) -> Self {
+        let buf = rmp_serde::to_vec(&val).unwrap();
         let mut deflater = DeflateEncoder::new(buf.as_slice(), Compression::best());
         let mut buf_c = Vec::new();
         deflater.read_to_end(&mut buf_c).unwrap();
         buf_c
     }
+}
 
-    pub fn from_bytes(
-        bytes: &[u8],
-    ) -> std::collections::BTreeMap<Witness, acir_field::FieldElement> {
+impl From<&[u8]> for WitnessMap {
+    fn from(bytes: &[u8]) -> Self {
         let mut deflater = DeflateDecoder::new(bytes);
         let mut buf_d = Vec::new();
         deflater.read_to_end(&mut buf_d).unwrap();
-        rmp_serde::from_slice(buf_d.as_slice()).unwrap()
+        Self(rmp_serde::from_slice(buf_d.as_slice()).unwrap())
+    }
+}
+
+impl From<WitnessMap> for Vec<acir_field::FieldElement> {
+    fn from(val: WitnessMap) -> Self {
+        val.0.into_values().collect()
     }
 }
```

Comment:
I'm a little cautious about implementing this as there isn't a single format for converting a `WitnessMap` into a `Vec<FieldElement>`.

e.g. when proving we fill in any missing indices with zeros: https://github.com/noir-lang/aztec_backend/blob/c9fb9e806f1400a2ff7594a0669bec56025220bb/src/acvm_interop/proof_system.rs#L78-L97 but when verifying we flatten the public inputs in the way you have here https://github.com/noir-lang/aztec_backend/blob/c9fb9e806f1400a2ff7594a0669bec56025220bb/src/acvm_interop/proof_system.rs#L66 

As there's not a single canonical conversion I don't think we can implement this trait.

Reply:
This was just the implementation needed for https://github.com/noir-lang/aztec_backend/pull/152/files#diff-38a8cdcfc939b53aa726b0f6a6a3c4940449409f27ffb409c36c64e4ea1b7749R65 since I didn't want to leak the `into_keys` or `into_values` functions which seem a little too specific to the internal data structure.

Reply:
Does `into_values` give us any more specific insight into the underlying implementation than this function though? They're the same thing but with a `.collect()`.

Reply:
It kind of implies there are 2 types of iterators; otherwise you'd just implement `Iter`. I thought about implementing `Iter` and giving it a tuple of `(Witness, FieldElement)`—would you want that instead?

Reply:
Yeah, I think returning an iterator over key-value pairs is the least opinionated iterator to give.

Reply:
Done, since it also needed to export the `IntoIter` struct, I moved it to a separate module. I'm in the process of updating the aztec_backend and Noir.